packages/loopring_v3/contracts/amm/AmplifiedAmmController.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+
+import "./IAmmController.sol";
+import "../amm/LoopringAmmPool.sol";
+import "../lib/Claimable.sol";
+import "../lib/MathUint.sol";
+import "../lib/MathUint96.sol";
+import "../thirdparty/SafeCast.sol";
+
+
+/// @author Brecht Devos - <brecht@loopring.org>
+contract AmplifiedAmmController is IAmmController, Claimable
+{
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+
+    uint public constant AMPLIFICATION_FACTOR_BASE = (10 ** 18);
+
+    mapping(address => uint) public amplificationFactors;
+
+    function getInitialVirtualBalances(
+        uint96[] memory joinAmounts
+        )
+        external
+        view
+        override
+        returns (uint96[] memory)
+    {
+        uint amplificationFactor = getAmplificationFactor(msg.sender);
+        uint96[] memory vTokenBalancesL2 = new uint96[](joinAmounts.length);
+        for (uint i = 0; i < joinAmounts.length; i++) {
+            vTokenBalancesL2[i] = (uint(joinAmounts[i]).mul(amplificationFactor) / AMPLIFICATION_FACTOR_BASE).toUint96();
+        }
+        return vTokenBalancesL2;
+    }
+
+    function getVirtualBalances(
+        uint96[] memory tokenBalancesL2,
+        uint96[] memory /*vTokenBalancesL2*/
+        )
+        external
+        view
+        override
+        returns (uint96[] memory)
+    {
+        // Only allow updating the virtual balances if the AF = 1
+        require(getAmplificationFactor(msg.sender) == AMPLIFICATION_FACTOR_BASE, "INVALID_OPERATION");
```

Comment:
Do you think there's a good reason to have a separate function? Would be 2 function calls instead of just one now so a bit more expensive, and making it fail in the function seems to have the same effect. Because the relayer needs to update all values as expected on-chain I also think it doesn't really help with making the transaction not fail because if not updated correctly the block will fail as well.

Reply:
When I ask this question, I was thinking there will be cases that changing virtual balances by the AMM owner is allowed (is it?), such as when one token's actual balance is 0. If that's the case, then having a method `allowSettingVirtualBalances` will make sense.

Reply:
The amplification factor allows us to settle orders using reduced slippage, but we don't have _have_ to make use of it. You can see it as an upper bound for the risk of LPs. If we want to settle orders with higher slippage we can start doing so without any actual changes to the pool parameters.

> I think the biggest concern I have is a pool becomes illiquid (the actual balance of one token is 0). I wish we have a way to reuse the pool by forcing all LPs to exit then reinitializing it using a fair market price.

But wouldn't that just be a new pool? And LPs of the previous pool may still want to remain in the old pool for when price moves within range again (and if they would have added their liquidity to the new pool they will have suffered some kind of extra IL).